### PR TITLE
update_UY_parser

### PR DESCRIPTION
**Pull request by mathilde-daugy**
## Description

When doing the DQR for 4Q 2022, i noticed that the UY parser was down and that the link we used was no longer available. I updated the parser to collect data from ADME (electricity market operator). This source has historical data for generation, consumption and exchanges.

Also I removed the part getting the Salto Grande data from Cammesa (AR) because it is included in the new source (see issue [1325
](https://github.com/electricitymaps/electricitymaps-contrib/issues/1325)

Data is available at 10-min intervals

Exchanges are also available but we use the values from the BR and AR parsers

### Preview
data for 01/01/2023

production
image

consumption
image

### CODE REVIEW

**Comment by VIKTORVAV99** [comment]
So eeh #4840 is already open for this 😅
However there are some issues with that PR that need to be addressed before merging so I'll compare the two and give my opinion on which should be closed or merged.

**Comment by mathilde-daugy** [comment]
> So eeh #4840 is already open for this 😅 However there are some issues with that PR that need to be addressed before merging so I'll compare the two and give my opinion on which should be closed or merged.
> 
> Initial looks favor this PR though as it only uses a single source (which was the cause of the issues in the other PR).

Ah my bad, i didn't see that...

> So eeh #4840 is already open for this 😅 However there are some issues with that PR that need to be addressed before merging so I'll compare the two and give my opinion on which should be closed or merged.
> 
> Initial looks favor this PR though as it only uses a single source (which was the cause of the issues in the other PR).

Ah my bad i didn't see that....
I also noticed that the other source doesn't enable us to backfill because it is a live data source. This means that if for some reason we can't get the data point, it is lost forever which is not ideal. 
I've checked and we can get data from at least 2020/01/01 using this source and also access spot prices.

**Comment by VIKTORVAV99** [comment]
> > So eeh #4840 is already open for this 😅 However there are some issues with that PR that need to be addressed before merging so I'll compare the two and give my opinion on which should be closed or merged.
> > Initial looks favor this PR though as it only uses a single source (which was the cause of the issues in the other PR).
> 
> Ah my bad i didn't see that.... I also noticed that the other source doesn't enable us to backfill because it is a live data source. This means that if for some reason we can't get the data point, it is lost forever which is not ideal. I've checked and we can get data from at least 2020/01/01 using this source and also access spot prices.

Yeah it looks like this source is way better quality wise, so don't worry to much about the other parser PR, it would most likely have been replaced by this sooner or later.

I'll try and do a full review tomorrow.

File: parsers/UY.py
Comment (unitrium) [inline parsers/UY.py:43]:
```suggestion
def get_adme_url(target_datetime: datetime, session: Session) -> str:
```
Since we are calling this function only from functions that handle the missing session case on their own, I would rather have this function just require a session.
This would add some clarity as the missing session problem would be treated all at the top and everything down just requires the session argument.

File: parsers/UY.py
Comment (unitrium) [inline parsers/UY.py:45]: No longer needed if the above is accepted.

File: parsers/UY.py
Comment (pierresegonne) [inline parsers/UY.py:16]: I think you can just do config.constants the full path does not work on windows if I'm not mistaken

File: parsers/UY.py
Comment (pierresegonne) [inline parsers/UY.py:96]: Isn't that a bit arbitrary ? 😅

File: parsers/UY.py
Comment (unitrium) [inline parsers/UY.py:155]: Shouldn't this be localized here? Because I guess their datasource is running on their timezone ?

File: parsers/UY.py
Comment (pierresegonne) [inline parsers/UY.py:182]: If we're fetching some exchanges with different parsers I'd suggest asserting that the given keys don't correspond to the exchange we fetch otherwise, or at least log it

File: parsers/UY.py
Comment (unitrium) [inline parsers/UY.py:54]:
```suggestion
    link = f"ADME_URL{date_format}&fecha_fin={next_day_format}&send=MOSTRAR"
```
It's probably a personal opinion, I tend to find f-strings more readable than concatenation.

File: parsers/UY.py
Comment (VIKTORVAV99) [reply parsers/UY.py:16]: Actually this one will work cause config is the only folder in there and wont work with config.constant as it's the only one that is not a symlink. (and config.constants would point to the other config folder and it don't have constants file).

It's all a little confusing 😅

File: parsers/UY.py
Comment (unitrium) [inline parsers/UY.py:47]:
```suggestion
    date_format = f"{target_datetime.day}%2F{target_datetime.month}%2F{target_datetime.year}"
```
You can also do this. It's up to you which one you prefer.

**Review by unitrium — APPROVED** [review]
Just a small question about the localization of the target_datetime otherwise it looks good to me!

File: parsers/UY.py
Comment (mathilde-daugy) [reply parsers/UY.py:96]: https://sunrise.maplogs.com/uruguay.120.html?year=2022
I had a look at that website for 2022 and thought it was an ok-ish range.
For some reason there is some solar production at night and it's really weird
image

File: parsers/UY.py
Comment (mathilde-daugy) [reply parsers/UY.py:122]: after renaming we have multiple hydro colulns

Comment (unitrium) [reply parsers/UY.py:122]: This will group production modes together. So for instance if you have COALITA and COALITO that are both mapped to COAL they will get summed together. It's a nice trick @mathilde-daugy  showed me on the ELEXON parser.

File: parsers/UY.py
Comment (mathilde-daugy) [reply parsers/UY.py:155]: I didn't localize here because i feel like it's not something that is done in other parser but i may be completely wrong

File: parsers/UY.py
Comment (VIKTORVAV99) [inline parsers/UY.py:178]: I think it would make sense to use this parser to fetch all or at least the AR exchange as this historical parsing support.
So we should probably update the exchange configs as well.

**Review by VIKTORVAV99 — COMMENTED** [review]
Pierre and Robin already covered most but I have some (nit-pick) suggestions as well. 😅

File: parsers/UY.py
Comment (mathilde-daugy) [reply parsers/UY.py:179]: I didn't realise that the exchanges from the AR parser were live. In this case, it's better to use this source.

File: parsers/UY.py
Comment (VIKTORVAV99) [reply parsers/UY.py:153]: That is usually dependant on the source data. Most "good" sources are fully iso compliant and use UTC time for the timestamps but this may not be the case here (I have not looked myself).

File: parsers/UY.py
Comment (mathilde-daugy) [reply parsers/UY.py:179]: I will also update the exchanges with BR because it also doesn't parse past dates

**Comment by VIKTORVAV99** [comment]
Before merging this I would like to add the author of #4840 as a contributor in the YAML file and then I'll go ahead an explain the situation and close said PR when this is merged.

**Comment by VIKTORVAV99** [comment]
> Before merging this I would like to add the author of #4840 as a contributor in the YAML file (and do feel free to add yourself as well if you want) and then I'll go ahead an explain the situation and close said PR when this is merged.

**Comment by mathilde-daugy** [comment]
> Before merging this I would like to add the author of #4840 as a contributor in the YAML file (do feel free to add yourself as well if you want) and then I'll go ahead an explain the situation and close said PR when this is merged.

No problem, please do :)

File: parsers/UY.py
Comment (VIKTORVAV99) [reply parsers/UY.py:94]: It's possible this also includes storage from solar generation sites such as battery storage but that's just a guess.

**Comment by VIKTORVAV99** [comment]
> > Before merging this I would like to add the author of #4840 as a contributor in the YAML file (do feel free to add yourself as well if you want) and then I'll go ahead an explain the situation and close said PR when this is merged.
> 
> No problem, please do :)

I added the contributors to the file as marked them as co-authors in my commit so feel free to merge this when it's ready.

File: parsers/UY.py
Comment (mathilde-daugy) [reply parsers/UY.py:94]: I would argue that we should set to 0 if we don't know ( the quantities are small anyways)

Comment (pierresegonne) [reply parsers/UY.py:94]: Could it be a timezone issue?

Comment (q--) [reply parsers/UY.py:94]: Solar could also include Concentrated Solar Power with energy storage. I found [this 2015 government study](http://www.energiasolar.gub.uy/index.php/investigacion-e-innovacion/investigacion-concentradores) investigating it, but haven't encountered anything they actually went through with it.

**Pull request merged**
